examples/protocol-and-stream-muxing/README.md

```diff
@@ -6,57 +6,76 @@ The feature of agreeing on a protocol over an established connection is what we
 
 After creating the nodes, we need to tell libp2p which protocols to handle.
 
 ```JavaScript
+const pipe = require('it-pipe')
+const { map } = require('streaming-iterables')
+const { toBuffer } = require('it-buffer')
+
 // ...
 const node1 = nodes[0]
 const node2 = nodes[1]
 
-// Here we are telling libp2p that is someone dials this node to talk with the `/your-protocol`
-// multicodec, the protocol identifier, please call this callback and give it the connection
+// Here we are telling libp2p that if someone dials this node to talk with the `/your-protocol`
+// multicodec, the protocol identifier, please call this handler and give it the stream
 // so that incomming data can be handled
-node2.handle('/your-protocol', (protocol, conn) => {
-  pull(
-    conn,
-    pull.map((v) => v.toString()),
-    pull.log()
+node2.handle('/your-protocol', ({ stream }) => {
+  pipe(
+    stream,
+    toBuffer,
+    map(String),
+    source => (async function () {
+      for await (const msg of source) {
+        console.log(msg)
+      }
+    })()
   )
 })
 ```
 
 After the protocol is _handled_, now we can dial to it.
 
 ```JavaScript
-node1.dialProtocol(node2.peerInfo, '/your-protocol', (err, conn) => {
-  if (err) { throw err }
-  pull(pull.values(['my own protocol, wow!']), conn)
-})
+const { stream } = await node1.dialProtocol(node2.peerInfo, ['/your-protocol'])
+
+await pipe(
+  ['my own protocol, wow!'],
+  stream
+)
 ```
 
 You might have seen this in the [Transports](../transports) examples. However, what it was not explained is that you can do more than exact string matching, for example, you can use semver.
 
 ```JavaScript
-node2.handle('/another-protocol/1.0.1', (protocol, conn) => {
-  pull(
-    conn,
-    pull.map((v) => v.toString()),
-    pull.log()
+node2.handle('/another-protocol/1.0.1', ({ stream }) => {
+  pipe(
+    stream,
+    toBuffer,
+    map(String),
+    source => (async function () {
+      for await (const msg of source) {
+        console.log(msg)
+      }
+    })()
   )
 })
 // ...
-node1.dialProtocol(node2.peerInfo, '/another-protocol/1.0.0', (err, conn) => {
-  if (err) { throw err }
-  pull(pull.values(['semver me please']), conn)
-})
+const { stream } = await node1.dialProtocol(node2.peerInfo, ['/another-protocol/1.0.0'])
+
+await pipe(
+  ['my own protocol, wow!'],
+  stream
+)
 ```
 
 This feature is super power for network protocols. It works in the same way as versioning your RPC/REST API, but for anything that goes in the wire. We had to use this feature to upgrade protocols within the IPFS Stack (i.e Bitswap) and we successfully managed to do so without any network splits.
 
 There is still one last feature, you can create your custom match functions.
 
+TODO: should we remove this?
```

Comment:
ideas ?

Reply:
Yes, remove it, we don't support custom match the same way anymore. The new method is to provide each of the protocols you want to handle for a specific handler. It may be valuable to replace it with an example of this. If a user has a backwards incompatible change, but it only requires a few changes to their code, they may wish to do protocol checking instead of having multiple handlers. An example might be a protobuf (or other) data structure having a breaking change. It might be easy to change the serializer being used and keep the remaining code.
```js
libp2p.handle(['/my-protocol/1.0.0', '/my-protocol/2.0.0'], myProtocolHandler)

function myProtocolHandler ({ connection, stream, protocol }) {
  if (protocol === '/my-protocol/2.0.0') {
    // handle backwards compatibility
  }
  ...
}
```

Reply:
I have an example with an array below, but I think it makes sense to mention that it can be used for `backwards incompatible change`